prov/rxm/src/rxm_ep.c

```diff
@@ -512,13 +516,28 @@ static ssize_t rxm_ep_send_common(struct fid_ep *ep_fid, const struct iovec *iov
 		pkt_size = sizeof(*pkt) + pkt->hdr.size;
 	}
 
-	ret = fi_send(rxm_conn->msg_ep, pkt, pkt_size, desc_tx_buf, 0, tx_entry);
-	if (ret) {
-		FI_WARN(&rxm_prov, FI_LOG_EP_DATA, "fi_send for MSG provider failed\n");
-		goto err;
+	if (flags & FI_INJECT) {
+		if (pkt_size <= rxm_ep->msg_info->tx_attr->inject_size) {
+			ret = fi_inject(rxm_conn->msg_ep, pkt, pkt_size, 0);
+			if (ret)
+				FI_WARN(&rxm_prov, FI_LOG_EP_DATA, "fi_inject for MSG provider failed\n");
+		} else {
+			FI_WARN(&rxm_prov, FI_LOG_EP_DATA, "passed data (size = %d) is too "
+				"big for MSG provider (max inject size = %d) \n",
+				pkt_size, rxm_ep->msg_info->tx_attr->inject_size);
+			ret = -FI_EINVAL;
```

Comment:
We should fall back to regular fi_send for the else case instead of reporting an error.